components/ai_chat/ai_chat_api.cc

```diff
@@ -113,40 +153,78 @@ void AIChatAPI::QueryPrompt(ResponseCallback callback,
+  return dict;
+}
 
-  DVLOG(1) << __func__ << " Prompt: " << prompt << "\n";
-  DVLOG(1) << __func__ << " Using model: " << model_name;
+void AIChatAPI::OnDataReceived(base::StringPiece string_piece,
+                               base::OnceClosure resume) {
+  std::vector<std::string> stream_data = base::SplitString(
+      string_piece, "\r\n", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  static constexpr char kDataPrefix[] = "data: {";
 
-  api_request_helper_.Request("POST", api_url, CreateJSONRequestBody(dict),
-                              "application/json", std::move(internal_callback),
-                              headers);
+  // Binding a remote will guarantee the order of the messages being sent, so we
+  // must ensure that it is connected.
+  DCHECK(data_decoder_);
+  DCHECK(json_parser_.is_connected());
 
-  DVLOG(1) << __func__ << " API Request sent\n";
+  for (const auto& data : stream_data) {
+    if (!base::StartsWith(data, kDataPrefix)) {
+      continue;
+    }
+
+    std::string json = data.substr(strlen(kDataPrefix) - 1);
+
+    data_decoder_->ParseJson(json,
+                             base::BindOnce(&AIChatAPI::OnParseJsonIsolated,
+                                            weak_ptr_factory_.GetWeakPtr()));
+  }
+
+  std::move(resume).Run();
 }
 
-void AIChatAPI::OnGetResponse(ResponseCallback callback,
-                              api_request_helper::APIRequestResult result) {
-  const bool success = result.response_code() == 200;
+void AIChatAPI::OnComplete(bool success) {
+  DCHECK(completion_callback_);
+  DCHECK(current_request_->get());
 
-  if (!success) {
-    DVLOG(1) << __func__ << " Response from API was not HTTP 200 (Received "
-             << result.response_code() << ")";
+  int response_code = -1;
+
+  if (network::SimpleURLLoader* simple_url_loader = current_request_->get();
+      simple_url_loader && simple_url_loader->ResponseInfo()->headers) {
+    response_code = simple_url_loader->ResponseInfo()->headers->response_code();
   }
 
-  std::string response = result.body();
-  const base::Value::Dict* dict = result.value_body().GetIfDict();
+  is_request_in_progress_ = false;
+  current_request_->reset(nullptr);
+  std::move(completion_callback_).Run(success, response_code);
+  api_request_helper_.Cancel(current_request_);
+}
 
-  if (dict) {
-    if (const std::string* completion = dict->FindString("completion")) {
-      response = *completion;
-    }
-  } else {
-    DVLOG(1) << __func__ << " Result dict not found\n";
+void AIChatAPI::OnRetry(base::OnceClosure start_retry) {
+  // Retries are not enabled for these requests.
+  NOTREACHED();
+}
+
+void AIChatAPI::OnParseJsonIsolated(
+    data_decoder::DataDecoder::ValueOrError result) {
+  if (!result.has_value()) {
+    return;
   }
 
-  std::move(callback).Run(response, success);
+  DCHECK(response_callback_);
+
+  if (const std::string* completion = result->FindStringKey("completion")) {
```

Comment:
The safe thing to do here is to drop the callback I'd say and debug why there's no completion key. 

The callback this function reports to is critical because on the other end it consumes the message which updates the UI.